[JSC] Implement `some`/`every`/`find` from Iterator Helpers Proposal

sosukesuzuki opened the pull request:
#### 08374f951b75d95ae15105d250c8ddbd23f2b50c
<pre>
[JSC] Implement `some`/`every`/`find` from Iterator Helpers Proposal
<a href="https://bugs.webkit.org/show_bug.cgi?id=279659">https://bugs.webkit.org/show_bug.cgi?id=279659</a>

Reviewed by Alexey Shvayka.

This patch implements following methods from Iterator Helpers Proposal[1]:

- `Iterator.prototype.some`[2]
- `Iterator.prototype.every`[3]
- `Iterator.prototype.find`[4]

[1]: <a href="https://github.com/tc39/proposal-iterator-helpers">https://github.com/tc39/proposal-iterator-helpers</a>
[2]: <a href="https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.some">https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.some</a>
[3]: <a href="https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.every">https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.every</a>
[4]: <a href="https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.find">https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.find</a>

* JSTests/stress/iterator-prototype-every.js: Added.
(assert):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameValue):
(sameValue.Iter.prototype.get next):
(sameValue.Iter):
(sameValue.gen):
(const.iter.next):
(Iterator.prototype.every.call.Iter.prototype.next):
(Iterator.prototype.every.call.Iter):
(Iterator.prototype.every.call):
(sameValue.Iter.prototype.next):
* JSTests/stress/iterator-prototype-find.js: Added.
(assert):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameValue):
(sameValue.Iter.prototype.get next):
(sameValue.Iter):
(sameValue.gen):
(const.iter.next):
(Iterator.prototype.find.call.Iter.prototype.next):
(Iterator.prototype.find.call.Iter):
(Iterator.prototype.find.call):
* JSTests/stress/iterator-prototype-some.js: Added.
(assert):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameValue):
(sameValue.Iter.prototype.get next):
(sameValue.Iter):
(sameValue.gen):
(const.iter.next):
(Iterator.prototype.some.call.Iter.prototype.next):
(Iterator.prototype.some.call.Iter):
(Iterator.prototype.some.call):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/builtins/JSIteratorPrototype.js: Added.
(some.wrapper.iterator):
(some):
(every.wrapper.iterator):
(every):
(find.wrapper.iterator):
(find):
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSIteratorPrototype::finishCreation):

Canonical link: <a href="https://commits.webkit.org/283677@main">https://commits.webkit.org/283677@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/df676178ead205bec6ee96c908ff4749941ad320

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67036 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46411 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19658 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71071 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18169 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54210 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17953 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/53762 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12226 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70103 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/42687 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58000 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34281 "Passed tests") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39358 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Running apply-patch; Checked out pull request; Skipped layout-tests; 4 new passes 7 flakes") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15394 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16523 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/60150 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61293 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15735 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72772 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66281 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10993 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15076 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61231 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11025 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58059 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61307 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14809 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9029 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2635 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/88049 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42218 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15495 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43295 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44478 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43038 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shvaikalesh reviewed (approved):
r=me but could you please run `Tools/Scripts/test262-runner -o test/built-ins/Iterator/prototype --save` and update the PR? Thank you.

sosukesuzuki committed:
[JSC] Implement `some`/`every`/`find` from Iterator Helpers Proposal
https://bugs.webkit.org/show_bug.cgi?id=279659

Reviewed by Alexey Shvayka.

This patch implements following methods from Iterator Helpers Proposal[1]:

- `Iterator.prototype.some`[2]
- `Iterator.prototype.every`[3]
- `Iterator.prototype.find`[4]

[1]: https://github.com/tc39/proposal-iterator-helpers
[2]: https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.some
[3]: https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.every
[4]: https://tc39.es/proposal-iterator-helpers/#sec-iteratorprototype.find

* JSTests/stress/iterator-prototype-every.js: Added.
(assert):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameValue):
(sameValue.Iter.prototype.get next):
(sameValue.Iter):
(sameValue.gen):
(const.iter.next):
(Iterator.prototype.every.call.Iter.prototype.next):
(Iterator.prototype.every.call.Iter):
(Iterator.prototype.every.call):
(sameValue.Iter.prototype.next):
* JSTests/stress/iterator-prototype-find.js: Added.
(assert):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameValue):
(sameValue.Iter.prototype.get next):
(sameValue.Iter):
(sameValue.gen):
(const.iter.next):
(Iterator.prototype.find.call.Iter.prototype.next):
(Iterator.prototype.find.call.Iter):
(Iterator.prototype.find.call):
* JSTests/stress/iterator-prototype-some.js: Added.
(assert):
(shouldThrow):
(throw.new.Error.Iter.prototype.next):
(throw.new.Error.Iter):
(throw.new.Error):
(sameValue):
(sameValue.Iter.prototype.get next):
(sameValue.Iter):
(sameValue.gen):
(const.iter.next):
(Iterator.prototype.some.call.Iter.prototype.next):
(Iterator.prototype.some.call.Iter):
(Iterator.prototype.some.call):
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/builtins/JSIteratorPrototype.js: Added.
(some.wrapper.iterator):
(some):
(every.wrapper.iterator):
(every):
(find.wrapper.iterator):
(find):
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSIteratorPrototype::finishCreation):

Canonical link: https://commits.webkit.org/283677@main
